### ISSUE TITLE
Unhandled socket error when making DNS query

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2492]
|<img alt="p1mrx's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| p1mrx reported|
|-|-|
|Trac ID|trac#2492|
|Type|defect|
|Created|2007-02-27 18:56:22Z|

A user of a program that I'm writing reported this error.  It occurs during "self.writeMessage(m, address)", and I think twisted should be catching this and calling the errback instead of letting the exception make its way back to my code.

```
Traceback (most recent call last):
 File "dtella.py", line 446, in <module>
 File "dtella.py", line 419, in run
 File "twisted\internet\posixbase.pyo", line 220, in run
 File "twisted\internet\posixbase.pyo", line 228, in mainLoop
--- <exception caught here> ---
 File "twisted\internet\base.pyo", line 561, in runUntilCurrent
 File "dtella_core.pyo", line 4198, in cb
 File "dtella.py", line 192, in startConnecting
 File "dtella_dnslookup.pyo", line 122, in getConfigFromDNS
 File "twisted\names\common.pyo", line 24, in query
 File "twisted\names\common.pyo", line 131, in lookupText
 File "twisted\names\client.pyo", line 313, in _lookup
 File "twisted\names\client.pyo", line 229, in queryUDP
 File "twisted\names\dns.pyo", line 1171, in query
 File "twisted\names\dns.pyo", line 1100, in writeMessage
 File "twisted\internet\udp.pyo", line 156, in write
socket.error: (10065, 'No route to host')
```

Attachments:

* [dns-query-except.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/de7/de71d0d4c989687b5589d77541ba52fe72d33099/aea2affa94763e38940a6008d1497d52c9ae557a.patch) (2155 bytes) - added by p1mrx on 2007-09-09 23:26:33Z - Proposed patch

<details><summary>Searchable metadata</summary>

```
trac-id__2492 2492
type__defect defect
reporter__p1mrx p1mrx
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1172602582000000 1172602582000000
changetime__1194020707000000 1194020707000000
version__None None
owner__ 
cc__therve cc__jml
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="p1mrx's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|p1mrx commented|
|-|-|

I already mentioned this on the mailing list, but a consequence of this bug is that, if writeMessage raises an exception, the value of 'resultDeferred' will be forgotten, causing a timeout to fire in the middle of nowhere a couple seconds later.  That definitely needs to be fixed.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [21264]) Apply patch, add tests.

Refs [#2492](https://github.com/twisted/twisted/issues/2492)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve removed owner|
|-|-|

Ready to review in query-write-errors-2492.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @therve|
|-|-|

I'd change the test_writeError docstring to say something like this:

''
Exceptions raised by the transport's write method should be turned into Failures passed to errbacks of the Deferred returned by L{<whichever protocol>.query}.
''

```_query```'s docstring should talk about the errors which the returned Deferred can have, I guess.

The rest looks good.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

(In [21378]) Some docs

Refs [#2492](https://github.com/twisted/twisted/issues/2492)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve removed owner|
|-|-|

Thanks, I added the docs above, back to review.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_9"></a>|@jml set owner to @therve|
|-|-|

There is a substantial amount of fixture set up in the two new `test_writeErrors`. That set up makes it hard to see which parts actually exercise the code.

I _think_ the best way to make this clearer is to extract two methods. The first method should create and return a protocol that uses the fake clock, the second should create and return a transport that raises an error on write.

Other than that, code looks good.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

(In [21398]) Add setUp methods to clean the tests.

Refs [#2492](https://github.com/twisted/twisted/issues/2492)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve removed owner|
|-|-|

I think I understood you concerns, but I tried to solve it using setUp, please tell me if it looks better. One of the good thing of this solution is that all the tests now use a deterministic clock, even if they don't explicitly manipulate it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @therve|
|-|-|

Looks good to me.  Nice simplifications to the existing tests.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve set status to `closed`|
|-|-|

(In [21535]) Merge query-write-errors-2492

Author: therve, p1mrx
Reviewer: exarkun
Fixes [#2492](https://github.com/twisted/twisted/issues/2492)

Handle write errors happening in dns queries, to have correct deferred
failures.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
